### PR TITLE
test/e2e/bench: Add bench test suite

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Based on #4247

Includes test for scenario from #4058

- Deploys Contour with resource limit to cluster and node selector for
contour nodes
- Deploys Envoys and test apps with node selector for app workload nodes
- Deploys apps + HTTPProxy sequentially, times time to successful
request as "ready"
- Outputs CSV of results and PNG scatter plot
- Exposes Envoy metrics port in daemonset and Contour metrics port to be used with `PodMonitor`
(see
    https://cloud.google.com/stackdriver/docs/solutions/gke/managing-metrics#configure-workload-metrics) for monitoring metrics
- Doesn't make any assertions on the results as of yet

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
have to increase the number of nodes to support the 4000 instance case, GKE has a limit of 110 pods per node

**Comment by sunjayBhatia** [comment]
also considering adding a warmup to ensure the initial opening on Envoy ports doesn't interfere with the actual data collection

**Comment by stevesloka** [comment]
> have to increase the number of nodes to support the 4000 instance case, GKE has a limit of 110 pods per node

Do you need pods to test? You could just create services + endpoints manually to create load, but I guess then you couldn't test the actual routing.

**Comment by sunjayBhatia** [comment]
> > have to increase the number of nodes to support the 4000 instance case, GKE has a limit of 110 pods per node
> 
> Do you need pods to test? You could just create services + endpoints manually to create load, but I guess then you couldn't test the actual routing.

yeah to mirror the knative test scenario from https://github.com/projectcontour/contour/issues/4058 we would need to measure how long it takes to an app being routable

a separate test could be to measure how long till HTTPProxy "valid" which wouldn't require actual app deployments

File: Makefile
Comment (skriss) [inline Makefile:337]: do we want to delete the cluster after running the test?

**Review by skriss — COMMENTED** [review]
This looks good, have you been able to run it to completion / do you have a sample result graph that you could share?

I can't remember off the top of my head if the knative test actually ran requests against the backends or just waited for some other indicator of the config being programmed (HTTPProxy status or checking Envoy config). Relatedly, since knative does scale to 0, I don't think that their test actually ended up keeping 4000 backend pods around on the cluster. May be something to keep an eye on.

**Comment by sunjayBhatia** [comment]
> have you been able to run it to completion / do you have a sample result graph that you could share?

yeah i can re-rerun to get the graph back, the full 4000 instance case is pretty slow/flaky :(

**Comment by sunjayBhatia** [comment]
> since knative does scale to 0, I don't think that their test actually ended up keeping 4000 backend pods around on the cluster

yeah thats a fair point, could kick off a scale down to 0/deployment deletion async once we get success on a particular instance, which would also reduce the endpoint processing/traffic etc.

**Comment by sunjayBhatia** [comment]
heres a partial graph, keep getting flakes and failures unfortunately, this one got to 860 instances

sequential-service-creation

results summary from ginkgo

```
      sequential-service-creation
      Name                     | N   | Min    | Median | Mean    | StdDev  | Max
      ===============================================================================
      time_to_ready [duration] | 860 | 36.6ms | 736ms  | 916.3ms | 743.2ms | 3.7352s
                               |     | echo-8 |        |         |         | echo-638
```

**Comment by sunjayBhatia** [comment]
doing some more experimenting, adding do not merge for a bit

**Comment by sunjayBhatia** [comment]
> I can't remember off the top of my head if the knative test actually ran requests against the backends or just waited for some other indicator of the config being programmed

I believe the knative "ready" status is set once their probing mechanism determines a backend is reachable

File: Makefile
Comment (sunjayBhatia) [reply Makefile]: added that, though with the makefile mechanics if youre relying on a generated name its sorta broken, you can pass a cluster name via env var though that should work

**Comment by youngnick** [comment]
I think the approach is looking great, but I think we should do the initial version with a number of pods that succeeds reliably somehow.

**Comment by sunjayBhatia** [comment]
> I think the approach is looking great, but I think we should do the initial version with a number of pods that succeeds reliably somehow.

running again to see what we can go with reliably

**Review by youngnick — APPROVED** [review]
LGTM as long as 1000 services seems reasonably stable. Let's not let the perfect be the enemy of the good.

File: Makefile
Comment (skriss) [inline Makefile:343]: Looks like we want L343 instead of L342 here so the test runs

File: test/scripts/gcp-bench-cluster.sh
Comment (skriss) [inline test/scripts/gcp-bench-cluster.sh:18]: what's this for?

**Review by skriss — CHANGES_REQUESTED** [review]
pretty much LGTM, couple minor things.

File: test/scripts/gcp-bench-cluster.sh
Comment (sunjayBhatia) [reply test/scripts/gcp-bench-cluster.sh:18]: previously i had two node pools to separate the worker nodes from contour nodes but when you add a node pool GKE always automatically repairs your cluster so its not usable really right after its created

this way you ask for some number of worker nodes and we add 1 which contour is deployed to (to separate contour and possibly allow us to add more specific resource constraints to test it in isolation, rather than it run on a node with other variable workloads)

Comment (skriss) [reply test/scripts/gcp-bench-cluster.sh:18]: Ah, makes sense, a short comment here would be helpful.

Comment (skriss) [reply test/scripts/gcp-bench-cluster.sh:23]: :+1:

**Review by skriss — APPROVED** [review]

**Pull request merged**
